pep-0718.rst

```diff
@@ -0,0 +1,152 @@
+PEP: 718
+Title: Subscriptable functions
+Author: James Hilton-Balfe <gobot1234yt@gmail.com>
+Sponsor: Guido van Rossum <guido@python.org>
+Discussions-To: https://discuss.python.org/t/26463/
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Content-Type: text/x-rst
+Created: 23-Jun-2023
+Python-Version: 3.13
+Post-History: `24-Jun-2023 <https://discuss.python.org/t/28457/>`__
+
+Abstract
+--------
+
+This PEP proposes making ``FunctionType``\ , ``MethodType``\ , ``BuiltinFunctionType``
+and ``BuiltinMethodType`` instances subscriptable for typing purposes. Doing so allows
+for specialisation in complex, hard-to-infer situations for type checkers where
+bi-directional inference (which allows for the types of parameters of anonymous
+functions to be inferred) is insufficient.
+
+Motivation
+----------
+
+Currently, it is not possible to infer certain situations for generic functions where
+unspecialised literals are involved
+
+.. code-block:: python
+
+   def make_list[T](*args: T) -> list[T]: ...
+   reveal_type(make_list())  # type checker cannot infer T
+
+Making instances of ``FunctionType`` subscriptable would allow for this constructor to
+be typed
+
+.. code-block:: python
+
+   reveal_type(make_list[int]())  # type is list[int]
+
+Currently you have to use an assignment to allow for this to be type-able
```

Comment:
Bidirectional type inference is yes a powerful tool for some of these situations but there are definitely cases where it can't be used as demonstrated in the examples. 

I don't think there's any problem with reveal type here, it is meant to show what the type checker sees.

The idea with cast whilst one line incurs more runtime cost than the alternative (which should be free) (it also requires call overhead compared to subscriptions the function directly) and it requires an import

Reply:
The objection isn't really about `reveal_type`, the objection is that in all of the examples we are calling a generic function and _not using the result_.  If we used the result, a checker could be expected to infer types based on the context.  And if we don't use the result, why do we really care what specialization is inferred?

I don't think the examples illustrate a compelling problem in practice.

Reply:
There are certainly usecases for not passing the result directly to a function where bi-directional inference can save us e.g.
```py
x = make_list()
x.insert(0, 1)
x[0].not_an_int_attr  # kaboom
```